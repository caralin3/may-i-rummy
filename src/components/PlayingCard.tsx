import { MaterialCommunityIcons } from '@expo/vector-icons';
import React from 'react';
import { Animated, Image, StyleSheet, View } from 'react-native';
import { Card, Colors } from 'react-native-paper';
import { Col, Row } from './Grid';
import { FontSizes, GutterSizes } from '../appearance';
import { Text } from './Text';
import { CardSuit, CardValue, Deck } from '../types';
import { createFlipAnim } from '../utils';

export interface PlayingCardFrontProps {
  suit: CardSuit;
  value: CardValue;
}

export const PlayingCardFront: React.FC<PlayingCardFrontProps> = ({ suit, value }) => {
  const cardFace = () => {
    const cardVal = Number(value);
    if (isNaN(cardVal)) {
      if (value === 'joker') {
        return (
          <>
            <Row>
              <Col flex={1}>{suitIcon(FontSizes.lg, 'club')}</Col>
              <Col flex={1}>{suitIcon(FontSizes.lg, 'diamond')}</Col>
            </Row>
            <Row justify='center'>
              <Col flex={1}>{suitIcon(FontSizes.lg, 'heart')}</Col>
              <Col flex={1}>{suitIcon(FontSizes.lg, 'club')}</Col>
            </Row>
          </>
        );
      }
      return <Row justify='center'>{suitIcon(60)}</Row>;
    } else {
      if (cardVal < 4) {
        let icons = [];
        for (let i = 1; i < cardVal + 1; i++) {
          icons.push(i);
        }
        return icons.map((icon) => (
          <Row key={icon} flex={1} justify='center'>
            {suitIcon()}
          </Row>
        ));
      } else if (cardVal > 3 && cardVal !== 8 && cardVal !== 10) {
        let icons: JSX.Element[] = [];
        icons.push(
          <Row key={0} flex={1} justify='center'>
            <Col flex={1}>{suitIcon()}</Col>
            <Col flex={1}>{suitIcon()}</Col>
          </Row>
        );
        let i = cardVal - 2;
        while (i > 0) {
          if (i % 2 === 0) {
            icons.push(
              <Row key={i} flex={1} justify='center'>
                <Col flex={1}>{suitIcon()}</Col>
                <Col flex={1}>{suitIcon()}</Col>
              </Row>
            );
            i--;
          } else {
            icons.push(
              <Row key={i} flex={1} justify='center'>
                <Col flex={1}>{suitIcon()}</Col>
              </Row>
            );
          }
          i--;
        }
        return icons;
      } else if (cardVal === 8) {
        let icons: JSX.Element[] = [];
        for (let j = 0; j < 5; j++) {
          if (j % 2 === 0) {
            icons.push(
              <Row key={j} flex={1} justify='center'>
                <Col flex={1}>{suitIcon()}</Col>
                <Col flex={1}>{suitIcon()}</Col>
              </Row>
            );
          } else {
            icons.push(
              <Row key={j} flex={1} justify='center'>
                <Col flex={1}>{suitIcon()}</Col>
              </Row>
            );
          }
        }
        return icons;
      } else if (cardVal === 10) {
        let icons: JSX.Element[] = [];
        let k = 0;
        while (k < 7) {
          if (k % 2 === 0) {
            icons.push(
              <Row key={k} flex={1} justify='center'>
                <Col flex={1}>{suitIcon()}</Col>
                <Col flex={1}>{suitIcon()}</Col>
              </Row>
            );
            if (k === 2) {
              k++;
            }
          } else {
            icons.push(
              <Row key={k} flex={1} justify='center'>
                <Col flex={1}>{suitIcon()}</Col>
              </Row>
            );
          }
          k++;
        }
        return icons;
      }
    }
    return <View style={styles.face}>{suitIcon()}</View>;
  };

  const suitIcon = (size = 20, cardSuit: CardSuit = suit) => (
    <MaterialCommunityIcons
      name={`cards-${cardSuit}`}
      color={cardSuit === 'heart' || cardSuit === 'diamond' ? Colors.redA700 : Colors.black}
      style={styles.suit}
      size={size}
    />
  );

  const cardValue = () => {
    const cardVal = Number(value);
    if (isNaN(cardVal)) {
      return value.charAt(0).toUpperCase();
    }
    return value;
  };

  return (
    <Card style={styles.container} elevation={5}>
      <Row flex={1} justify='center'>
        {value === 'joker' ? (
          <Col style={styles.joker}>
            <Text>
              J{'\n'}O{'\n'}K{'\n'}E{'\n'}R
            </Text>
          </Col>
        ) : (
          <Col style={styles.value}>
            <Text>{cardValue()}</Text>
            {suitIcon(FontSizes.md)}
          </Col>
        )}
        <View style={styles.numbers}>{cardFace()}</View>
      </Row>
    </Card>
  );
};

export interface PlayingCardBackProps {
  deck: Deck;
}

export const PlayingCardBack: React.FC<PlayingCardBackProps> = ({ deck }) => {
  const backPattern =
    deck === 'blue'
      ? require('../appearance/assets/card-back-blue.png')
      : require('../appearance/assets/card-back-red.png');

  return (
    <Card style={styles.container} elevation={5}>
      <Image source={backPattern} resizeMode='contain' style={styles.backFace} />
    </Card>
  );
};

export interface PlayingCardProps {
  deck: Deck;
  faceDown?: boolean;
  suit: CardSuit;
  value: CardValue;
}

export const PlayingCard: React.FC<PlayingCardProps> = (props) => {
  const flipAnim = createFlipAnim(new Animated.Value(0));

  // React.useEffect(() => {
  //   flipAnim.flipCard();
  // }, []);

  return (
    <View>
      <Animated.View style={[styles.flip, flipAnim.frontAnimatedStyle]}>
        {props.faceDown ? <PlayingCardBack {...props} /> : <PlayingCardFront {...props} />}
      </Animated.View>
      <Animated.View style={[styles.flip, styles.flipBack, flipAnim.backAnimatedStyle]}>
        {props.faceDown ? <PlayingCardFront {...props} /> : <PlayingCardBack {...props} />}
      </Animated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    borderRadius: 10,
    height: 150,
    marginBottom: 10, //remove
    padding: GutterSizes.xxs,
    width: 100,
  },
  backFace: {
    flex: 1,
    marginVertical: GutterSizes.xxs,
    width: '100%',
  },
  numbers: {
    alignItems: 'center',
    flex: 1,
    paddingHorizontal: 20,
    paddingVertical: 15,
    width: '100%',
  },
  face: {
    alignItems: 'center',
  },
  flip: {
    backfaceVisibility: 'hidden',
  },
  flipBack: {
    position: 'absolute',
    top: 0,
  },
  joker: {
    left: GutterSizes.xs,
    position: 'absolute',
    top: GutterSizes.xs,
  },
  suit: {
    textAlign: 'center',
  },
  value: {
    left: 0,
    position: 'absolute',
    top: GutterSizes.xs,
  },
});
