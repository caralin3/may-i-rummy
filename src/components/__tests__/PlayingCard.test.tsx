import React from 'react';
import renderer from 'react-test-renderer';
import { PlayingCard } from '../PlayingCard';

describe('Playing Card', () => {
  test('renders correctly', () => {
    const wrapper = renderer
      .create(
        <PlayingCard
          card={{
            deck: 'blue',
            suit: 'club',
            value: '6',
          }}
        />
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
  test('renders face down correctly', () => {
    const wrapper = renderer
      .create(
        <PlayingCard
          faceDown
          card={{
            deck: 'blue',
            suit: 'club',
            value: '6',
          }}
        />
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
