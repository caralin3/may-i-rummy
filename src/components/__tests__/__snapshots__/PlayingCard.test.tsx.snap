// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playing Card renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backfaceVisibility": "hidden",
        "transform": Array [
          Object {
            "rotateX": "0deg",
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "rgba(0, 0, 0, 0.12)",
          "borderRadius": 10,
          "elevation": 5,
          "height": 150,
          "marginBottom": 10,
          "padding": 3,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 5,
          "width": 100,
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "width": "100%",
              },
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
                "paddingHorizontal": undefined,
                "paddingVertical": undefined,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": undefined,
                "justifyContent": "center",
                "left": 0,
                "padding": undefined,
                "position": "absolute",
                "top": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": undefined,
                  "fontSize": 24,
                  "fontStyle": undefined,
                  "fontWeight": undefined,
                  "lineHeight": 24,
                  "paddingBottom": undefined,
                  "paddingLeft": undefined,
                  "paddingRight": undefined,
                  "paddingTop": undefined,
                  "textAlign": undefined,
                }
              }
            >
              6
            </Text>
            <Text />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 15,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "width": "100%",
                  },
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "paddingHorizontal": undefined,
                    "paddingVertical": undefined,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "width": "100%",
                  },
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "paddingHorizontal": undefined,
                    "paddingVertical": undefined,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "width": "100%",
                  },
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "paddingHorizontal": undefined,
                    "paddingVertical": undefined,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backfaceVisibility": "hidden",
        "position": "absolute",
        "top": 0,
        "transform": Array [
          Object {
            "rotateX": "180deg",
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "rgba(0, 0, 0, 0.12)",
          "borderRadius": 10,
          "elevation": 5,
          "height": 150,
          "marginBottom": 10,
          "padding": 3,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 5,
          "width": 100,
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <Image
          index={0}
          resizeMode="contain"
          siblings={
            Array [
              "Image",
            ]
          }
          source={1}
          style={
            Object {
              "flex": 1,
              "marginVertical": 3,
              "width": "100%",
            }
          }
          total={1}
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Playing Card renders face down correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backfaceVisibility": "hidden",
        "transform": Array [
          Object {
            "rotateX": "0deg",
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "rgba(0, 0, 0, 0.12)",
          "borderRadius": 10,
          "elevation": 5,
          "height": 150,
          "marginBottom": 10,
          "padding": 3,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 5,
          "width": 100,
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <Image
          index={0}
          resizeMode="contain"
          siblings={
            Array [
              "Image",
            ]
          }
          source={1}
          style={
            Object {
              "flex": 1,
              "marginVertical": 3,
              "width": "100%",
            }
          }
          total={1}
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backfaceVisibility": "hidden",
        "position": "absolute",
        "top": 0,
        "transform": Array [
          Object {
            "rotateX": "180deg",
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "rgba(0, 0, 0, 0.12)",
          "borderRadius": 10,
          "elevation": 5,
          "height": 150,
          "marginBottom": 10,
          "padding": 3,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 5,
          "width": 100,
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "width": "100%",
              },
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
                "paddingHorizontal": undefined,
                "paddingVertical": undefined,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": undefined,
                "justifyContent": "center",
                "left": 0,
                "padding": undefined,
                "position": "absolute",
                "top": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": undefined,
                  "fontSize": 24,
                  "fontStyle": undefined,
                  "fontWeight": undefined,
                  "lineHeight": 24,
                  "paddingBottom": undefined,
                  "paddingLeft": undefined,
                  "paddingRight": undefined,
                  "paddingTop": undefined,
                  "textAlign": undefined,
                }
              }
            >
              6
            </Text>
            <Text />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 15,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "width": "100%",
                  },
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "paddingHorizontal": undefined,
                    "paddingVertical": undefined,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "width": "100%",
                  },
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "paddingHorizontal": undefined,
                    "paddingVertical": undefined,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "width": "100%",
                  },
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "paddingHorizontal": undefined,
                    "paddingVertical": undefined,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "padding": undefined,
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
